### PR TITLE
[part of #832] use db_id between projects and roles

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

✔️ `iam_project_roles` now refers to `iam_projects` using the latter's `db_id` field, not `id`.

### :chains: Related Resources

#832 

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/storage/v2/postgres/postgres.go [hunk 1]
```diff
@@ -888,14 +882,8 @@ func checkIfRoleIntersectsProjectsFilter(ctx context.Context, q Querier,
...
-	// doesn't exist, this will return 0 rows which will bubble up to NotFoundErr when passed to processError.
-	row := q.QueryRowContext(ctx,
-		`SELECT COALESCE(array_agg(rp.project_id)
-				FILTER (WHERE rp.project_id IS NOT NULL), '{(unassigned)}') && $2 AS intersection
-			FROM iam_roles AS r
-			LEFT OUTER JOIN iam_role_projects AS rp ON rp.role_id=r.db_id
-			WHERE r.id = $1;`,
-		id, pq.Array(projectsFilter))
+	// doesn't exist, this will return a proper SQL "no rows" error when passed to processError.
+	row := q.QueryRowContext(ctx, "SELECT projects_match(role_projects($1), $2)", id, pq.Array(projectsFilter))
```
Comment (srenatus) [inline components/authz-service/storage/v2/postgres/postgres.go]: 🤔 would make sense to coalesce this into `{(unassigned)}`, but the tests pass regardless...

File: components/authz-service/storage/v2/postgres/postgres.go [hunk 2]
```diff
@@ -1372,7 +1358,7 @@ func (p *pg) CreateProject(ctx context.Context, project *v2.Project) (*v2.Projec
 	}
 
 	if project.Type == v2.Custom {
-		row := tx.QueryRowContext(ctx, ` WITH t as (SELECT id from iam_projects WHERE type='custom') SELECT COUNT(*) FROM t;`)
+		row := tx.QueryRowContext(ctx, "SELECT count(*) FROM iam_projects WHERE type='custom'")
```
Comment (srenatus) [inline components/authz-service/storage/v2/postgres/postgres.go]: drive-by cleanup

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 3]
```diff
@@ -40,4 +40,18 @@ CREATE OR REPLACE FUNCTION
 
 $$ LANGUAGE sql;
 
+CREATE FUNCTION
+  role_projects(_role_id TEXT)
+  RETURNS TEXT[] AS $$
+
+  SELECT COALESCE(array_agg(p.id) FILTER (WHERE p.id IS NOT NULL), '{(unassigned)}')
```
Comment (bcmdarroch) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: nifty 😋

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  BEGIN
+      SELECT db_id INTO STRICT role_db_id
+      FROM iam_roles AS r
+      WHERE r.id=_role_id;
+      IF FOUND THEN
+          SELECT array_agg(p.id) INTO _project_ids
+          FROM iam_role_projects AS rp
+          LEFT JOIN iam_projects AS p
+          ON rp.project_id=p.db_id
+          WHERE rp.role_id=role_db_id;
```
Comment (tylercloke) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: [nit] Some tabbing might make this a bit more readable.

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  role_projects(_role_id TEXT, OUT _project_ids TEXT[])
+  RETURNS TEXT[] AS $$
+
+  DECLARE
+  role_db_id INTEGER;
+  BEGIN
+      SELECT db_id INTO STRICT role_db_id
+      FROM iam_roles AS r
+      WHERE r.id=_role_id;
+      IF FOUND THEN
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: I am torn... a successful find sets `FOUND`, yes, but an unsuccessful find throws an exception. So you do not really need the `IF FOUND...`; it will work exactly the same without it.

However... on the one hand, it is more clear to have `IF FOUND...` because you know that's what the author really meant. But on the other hand, it makes one think (incorrectly) that a 'not found' case just falls through.
🤔 Not sure what is "better". What do you think about keeping the `IF FOUND` and just above the `END IF` add a comment `-- ELSE throw SQLException` ?

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+          SELECT array_agg(p.id) INTO _project_ids
+          FROM iam_role_projects AS rp
+          LEFT JOIN iam_projects AS p
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: `LEFT JOIN` implies there is the possibility that there will be some null records in rp.project_id that does not have a corresponding p.db_id, but that should not be possible, so `INNER JOIN` would be more appropriate.

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      'actions', r.actions,
+      'projects', COALESCE(role_projects(r.id), '{}')
+    ) AS role
+  FROM iam_roles AS r
+  WHERE projects_match(COALESCE(role_projects(r.id), '{(unassigned)}'),  _projects_filter);
+
+$$ LANGUAGE sql;
+
+-- helper functions, now made strict: if they don't find anything,
+-- a "not found" exception is thrown
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: Should we be more complete with the comment? To the effect of "must match exactly one row; throws an exception if none (not found) or more than one (too many rows) are returned"

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            iam_projects
+        WHERE
+            id = t.project_temp_id);
+ALTER TABLE iam_role_projects
+    DROP COLUMN project_temp_id;
+
+-- returns NULL if there's no associated projects; the call can decide
+-- (by coalesce'ing) if they want that to be {} or {(unassigned)}.
+CREATE OR REPLACE FUNCTION
+  role_projects(_role_id TEXT, OUT _project_ids TEXT[])
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: Is there an advantage to having the `OUT` parameter instead of just returning a value?

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql (hunk 5 above)
Comment (srenatus) [reply components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: I've been thinking more about this one, too. At some point, I wanted it to return `{(unassigned)}`, but then I've reconsidered. And I think you're right, if we don't do that, we don't need the `IF FOUND` conditional. 👍

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql (hunk 8 above)
Comment (srenatus) [reply components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: We cannot `SELECT INTO` without a named parameter... we might not _strictly have to_ do this for the projects part of this, but I don't suppose it hurts 😉

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+          LEFT JOIN iam_projects AS p
+          ON rp.project_id=p.db_id
+          WHERE rp.role_id=role_db_id;
+     END IF;
+  END;
+$$ LANGUAGE plpgsql;
+
+
+CREATE OR REPLACE FUNCTION
+  query_role(_role_id TEXT)
```
Comment (msorens) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: Just noticed: shouldn't this function also have the second parameter of `_projects_filter` ?

Comment (srenatus) [reply components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: If we wanted to unify how project filtering works, yes. However, I've considered that out of scope for this issue. Furthermore, I'm not convinced about what the best approach is just yet.

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql (hunk 4 above)
Comment (srenatus) [reply components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: I've given all my trust to `pg_format` 😉

**Comment by srenatus** [comment]
ℹ️ will merge when green

File: components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    SELECT
+        db_id INTO STRICT _db_id
+    FROM
+        iam_projects
+    WHERE
+        id = _id;
+END;
+$$
+LANGUAGE plpgsql;
+CREATE OR REPLACE FUNCTION project_id (_db_id iam_projects.db_id % TYPE, OUT _id iam_projects.id % TYPE)
```
Comment (blakestier) [inline components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: Are we using this anywhere? I can't seem to find it 🤔

Comment (msorens) [reply components/authz-service/storage/postgres/migration/sql/55_use_project_db_id_in_references_of_project_roles.up.sql]: I agree; don't see it being used anywhere. 🤷‍♂
